index/segment/mem/segment.go

```diff
@@ -93,26 +97,60 @@ func NewSegment(offset postings.ID, opts Options) (sgmt.MutableSegment, error) {
 
 func (s *segment) Insert(d doc.Document) error {
 	s.state.RLock()
+	defer s.state.RUnlock()
 	if s.state.closed {
-		s.state.RUnlock()
 		return sgmt.ErrClosed
 	}
 
 	if s.state.sealed {
-		s.state.RUnlock()
 		return errSegmentSealed
 	}
 
-	// Validate that the document contains only valid UTF-8.
+	// Check if the document contains a document ID and ensure it's fields are valid UTF-8.
+	var docID []byte
 	for _, f := range d.Fields {
+		// TODO: Should we enforce uniqueness of field names?
 		if !utf8.Valid(f.Name) {
 			return fmt.Errorf("document contains invalid field name: %v", f.Name)
 		}
+
+		if bytes.Equal(f.Name, doc.IDFieldName) {
+			if docID != nil {
+				return errMultipleIDs
+			}
+
+			if len(docID) == 0 {
+				return errZeroLengthID
+			}
+
+			docID = f.Value
+		}
+
 		if !utf8.Valid(f.Value) {
 			return fmt.Errorf("document contains invalid field value: %v", f.Value)
 		}
 	}
 
+	if docID != nil {
+		pl, err := s.matchTerm(doc.IDFieldName, docID)
```

Comment:
couple of nits about this: 
- don't need to grab the lock in matchTerm for this codepath
- don't need to allocate a postings list for this case either (termsDict.matchTerm) does so

Could you create a separate code path which is lighter weight (`segment.containsTermWithLock` and `termsDict.containsField`)?

Reply:
sounds good